### PR TITLE
DR-2831 Add support for predictable and global DRS IDs

### PR DESCRIPTION
**Pull request by nmalfroy**
Woof...apologies for the size of this PR.  In fairness, a lot of this is testing.  This is really probably 2 PRs.  The first commit and then what's after.

## Intro
This PR aims to solve the issue where a DRS ID really should be able to point to files that exist across multiple regions and multiple clouds.  This is very common in NIH projects and is needed for TDR to be successful in replacing existing solutions there.

The issue is that TDR has a 1:1 Dataset to Snapshot relationship and in order to have multiple regions and clouds, you need different datasets which, using our current DRS ID scheme, won't work since it includes the snapshot id.

The solution implemented in this PR is to introduce:
- Predictable File Ids
- Global Drs Ids

Both are independently useful features but together they allow us to support the NIH use-case.

I'll note one nice side-effect feature with the Global Drs Ids: we can now have a ton of snapshots created from the same Dataset and each snapshot can have the same Drs Ids.  This has been a source of consternation in the past.

More information can be found here: https://docs.google.com/document/d/1i7Nk99fQWZBVeOAvqqTymT1oG27r6koRUnEsE2mYKKA/edit#heading=h.6woj4vghhmz3

## Predictable File IDs
First commit:
- Adds support for predictable file ids (a behavior that is enabled when you create the dataset)
- If false, fileIds continue to be minted randomly the way that it works today
- If true, fileIds are creates as a function of targetPath + size + MD5 hash
  - With this approach, one thing that's fun is that we need to look up the source file before copying it to get its MD5 hash (since we use the fileId for the target location of the file)
  - We also need to swap the order of the steps where we do the file copy and record the directory entry for each file (since the file copy step is the one where we get the file id in the predictable file id case)
  - Note: directories will not have predictable ids. Not sure there's a need and it's a looooot more expensive to calculate

## Global Drs IDs
Following commits:
- Adds support for a new type of Drs ID.  Where before DRS IDs look like `v1_<snapshotuuid>_<fileuuid>`, this PR introduces a new type that is global that looks like `v2_fileuuid`.
- It's enabled via a new (and immutable) Snapshot parameter: `globalFileIds`
- To support this, a new table named `drs_id` is added and it's job is to store a lookup from this drs id to all snapshots that contain it
- A note on perf: I tested with ~10M records in the table the performance of the lookup adds negligible overhead compared to Firestore lookups
- The snapshot creation flight now takes care of:
  - Recording the drs ids
  - Making the views/parquet files show the new drs URI in that case (e.g. `drs://<hostname>/v2_fileuuid`)
- The DrsService changed over its logic (this is where a lot of the code changes are)
  - Snapshots associated with a drs ID are now assumed to be a list, not just a single value
  - The DrsObject lookup looks up the Drs object information for all snapshots that the user has access to and then merges the information.
    - The merging is somewhat straightforward: it does a lookup and assumes single values for everything (and will fail if there are multiple values) except:
      - aliases
      - checksums: it merges on checksum types and fails if there are overlapping checksums that don't match (e.g. if there are 2 md5 hashes that don't match)
      - accessMethods: this is the trickier one. For this, unions all of the access methods but the access Id now has `*<billing snapshot uuid>` appended to it.  The reason that this is there is because in order to access a requester pays bucket, we need to specify a snapshot whose project can be billed for the access.  The algorithm is to group all snapshots by billing id and then use the first (by ID sort order) snapshot in the group (since the snapshot in question doesn't matter, if they all point to the same billing account)

I'll note (and I mentioned in standup) that I haven't added any integration tests here.  Everything is tested via unit tests (a little bit of connected testing for Synapse stuff).  I may still add an integration test for sanity but the real guts of the logic is exercised via unit tests.  So that's good :)

### CODE REVIEW

File: src/main/java/bio/terra/service/dataset/Dataset.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/dataset/Dataset.java]: Naming nit: maybe makes more sense as "hasPredictableFileIds" or "usesPredictableFileIds".

File: src/main/java/bio/terra/service/dataset/Dataset.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/dataset/Dataset.java]: Nit: I think we like to follow Google's Java style guide, and would camel case the parameter name as `predictableFileIds`.  https://google.github.io/styleguide/javaguide.html#s5.3-camel-case

File: src/main/java/bio/terra/service/dataset/DatasetDao.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/dataset/DatasetDao.java:390]: Hooray for multi-line strings!  The indentation of the new lines is slightly off though, if you wouldn't mind fixing.

File: src/main/java/bio/terra/service/filedata/flight/ingest/FileIngestFlight.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/filedata/flight/ingest/FileIngestFlight.java]: I know you mentioned in person why we needed to reverse the order of these steps when generating predictable file IDs, but I think it would be worth a comment in code too!  A supplementary change would be to define the steps once outside of the block, then add their references in the appropriate order.

File: src/main/java/bio/terra/service/filedata/flight/ingest/FileIngestFlight.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/filedata/flight/ingest/FileIngestFlight.java]: Same feedback here.

File: src/main/java/bio/terra/service/filedata/flight/ingest/FileIngestWorkerFlight.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/filedata/flight/ingest/FileIngestWorkerFlight.java]: And… if I had scrolled down I would have seen these comments! :) 

Given the proliferation of this pattern across two flights and two platforms… I wonder if there's an opportunity to break this out into a shared utility?

File: src/main/java/bio/terra/service/filedata/flight/ingest/IngestFileAzurePrimaryDataStep.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/filedata/flight/ingest/IngestFileAzurePrimaryDataStep.java]: Is there an advantage to this conditional wrapper?  Let's say the dataset uses the new predictable IDs.  If so, at this point the working map wouldn't have a `FILE_ID`, so this `workingMap.get` would return a null value anyway, right?

File: src/main/java/bio/terra/service/filedata/flight/ingest/IngestFileAzurePrimaryDataStep.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/filedata/flight/ingest/IngestFileAzurePrimaryDataStep.java:82]: Similar question about the conditional's value here.  If `fileId` is non-null it's because the dataset had the old-style IDs, so `fsFileInfo` would use the same one in the map… it would write to the map what's already there, which could make things a bit more readable.

File: src/main/resources/api/data-repository-openapi.yaml
Comment (okotsopoulos) [inline src/main/resources/api/data-repository-openapi.yaml:3726]: In these cases with repeated column definitions and descriptions, I like to break them out into reusable references in case we end up changing them in the future… but I know that means that the default value won't propagate for free through the controller!

File: src/main/java/bio/terra/service/filedata/DrsIdDao.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/filedata/DrsIdDao.java:18]: This may be a matter of personal preference, but when I see query strings with parameter references I'd prefer them to be defined alongside the creation of the parameter source.  Thoughts?

File: src/test/java/bio/terra/service/snapshot/SnapshotDaoTest.java
Comment (okotsopoulos) [inline src/test/java/bio/terra/service/snapshot/SnapshotDaoTest.java:843]: Another vote for getting to that JUnit 5 migration… :)

File: src/main/java/bio/terra/service/filedata/azure/AzureSynapsePdao.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/filedata/azure/AzureSynapsePdao.java]: Would you instead want to leverage the `logQuery` helper method?

File: src/main/java/bio/terra/service/snapshot/Snapshot.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/snapshot/Snapshot.java]: Similar naming feedback here, would prefer "hasGlobal…" or "usesGlobal…".

File: src/main/java/bio/terra/service/snapshot/SnapshotDao.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/snapshot/SnapshotDao.java:655]: Any particular reasoning for the breakdown of lines like this?

File: src/main/java/bio/terra/service/snapshot/flight/create/SnapshotRecordFileIdsAzureStep.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/snapshot/flight/create/SnapshotRecordFileIdsAzureStep.java:17]: I like this pattern of sharing step code when we need slightly different implementations by platform.

File: src/main/java/bio/terra/service/filedata/DrsService.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/filedata/DrsService.java:801]: This may be more trouble than it's worth, but it looks like `mergeDRSObjects` will throw on the first non-unique value of our singletons.  Any way to collect those conflicts and throw them all at the end?

File: src/test/java/bio/terra/service/filedata/DrsServiceTest.java
Comment (okotsopoulos) [inline src/test/java/bio/terra/service/filedata/DrsServiceTest.java]: I think this method name may be a typo!

More meaningfully, it would be great to also see tests demonstrating / verifying behavior when 1+ required unique fields is not unique.

File: src/main/java/bio/terra/service/filedata/DrsService.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/filedata/DrsService.java]: Would you mind adding unit test(s) for this method as well?

File: src/main/java/bio/terra/service/filedata/DrsService.java
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/filedata/DrsService.java]: Food for thought, it might be easier to do more targeted and complete unit testing of this method if these stream operations were broken out into their own methods and then tested individually.  That could also sidestep the need for the comments and make the merge more readable. Example:

```
…
  .createdTime(getEarliestCreatedTime(drsObjects))
…

@VisibleForTesting
String getEarliestCreatedTime(List<DRSObject> drsObjects) {
  return drsObjects.stream()
                    .map(DRSObject::getCreatedTime)
                    .map(Instant::parse)
                    .min(Comparator.naturalOrder())
                    .map(Instant::toString)
                    .orElse("");
}
```

File: src/test/java/bio/terra/service/filedata/DrsServiceTest.java
Comment (okotsopoulos) [inline src/test/java/bio/terra/service/filedata/DrsServiceTest.java:233]: Another piece of food for thought: this unit test does a lot of heavy lifting.  In my own testing I've found this to be a frequent indicator that the underlying code could benefit from further breakdown or reorganization, which makes the subcomponents easier to test.

**Review by okotsopoulos — COMMENTED** [review]
Thank you so much for breaking this PR into manageable commits, that and the description made this a lot less daunting to review!

I left a high volume of comments so my apologies for that, most of them are either small nits or philosophical musings on testing approach :-)

I'm holding off approving to allow for conversation, but can review again on request.

File: src/main/java/bio/terra/service/dataset/DatasetSummary.java
Comment (snf2ye) [inline src/main/java/bio/terra/service/dataset/DatasetSummary.java]: Why do we need to put this feature behind a flag on the dataset? Could we just do this for all file ingests going forward? That could mean that we have datasets with a mix of random file ids and predictable file ids, but I'm not sure why that would matter.

File: src/main/java/bio/terra/service/filedata/FileIdService.java
Comment (snf2ye) [inline src/main/java/bio/terra/service/filedata/FileIdService.java]: While it's probably still very unlikely we would run into a hash collision, what would happen if we ended up with two different files with same UUID? 
I haven't totally parsed through the possible scenarios, but I was looking at IngestFileDirectoryStep, and I think if we happened to have a file with the same directory, loadTag, and, now, fileId, we would overwrite the existing file.

File: src/main/java/bio/terra/service/filedata/azure/AzureSynapsePdao.java
Comment (snf2ye) [reply src/main/java/bio/terra/service/filedata/azure/AzureSynapsePdao.java]: I'm betting this was just leftover from testing - I don't think we should log the query.  
Can we get rid of the logQuery() method, too?

**Review by snf2ye — COMMENTED** [review]
You know it's a hefty PR when the browser struggles to load the diffs 😆 

I want to spend a little more time understanding the DRSService changes/the new mergeDRSObjects object, but otherwise looks great - this is an exciting change!

File: src/main/java/bio/terra/service/filedata/azure/AzureSynapsePdao.java
Comment (nmalfroy) [reply src/main/java/bio/terra/service/filedata/azure/AzureSynapsePdao.java]: Yeah, though I'm going to leave the logging (we do it for BQ queries as well).  It can be super useful for debugging

Comment (nmalfroy) [reply src/main/java/bio/terra/service/filedata/azure/AzureSynapsePdao.java]: I'll remove this particular statement

File: src/main/java/bio/terra/service/filedata/FileIdService.java
Comment (nmalfroy) [reply src/main/java/bio/terra/service/filedata/FileIdService.java]: Right, so the assumption is that if there is a collision, you would only see the file that you have access to (based on the snapshot) so collisions aren't a problem from a security point of view.  Where it might get interesting is if a user has access to both but there's not really a good way to check for collisions and I'm not too concerned given the low probability.  The MD5 (prob = 1.47*10^29) and the path and the size have to match...that probability is exceedingly small

File: src/main/java/bio/terra/service/dataset/DatasetSummary.java
Comment (nmalfroy) [reply src/main/java/bio/terra/service/dataset/DatasetSummary.java]: That's a great question.  Mostly, I was thinking that this gives us a chance to ease into this feature

File: src/test/java/bio/terra/service/filedata/DrsServiceTest.java
Comment (nmalfroy) [reply src/test/java/bio/terra/service/filedata/DrsServiceTest.java:233]: Yeah, I was torn about testing granularity.  I kinda like having a test like this to demonstrate the full walkthrough of the lookup and having the various edge cases (advantage of the unit tests...they're super fast!).  But I can certainly break apart the methods some more like for the date bit that gets earliest and latest

File: src/main/java/bio/terra/service/filedata/DrsService.java
Comment (nmalfroy) [reply src/main/java/bio/terra/service/filedata/DrsService.java]: totally!

Comment (nmalfroy) [reply src/main/java/bio/terra/service/filedata/DrsService.java]: (though incidentally, my thought was that this got exercised by some of the slightly bigger unit test)

File: src/test/java/bio/terra/service/filedata/DrsServiceTest.java
Comment (nmalfroy) [reply src/test/java/bio/terra/service/filedata/DrsServiceTest.java]: derp...it totally is. I fixed it and forgot to check it in 🤦‍♂️

File: src/main/java/bio/terra/service/filedata/DrsService.java
Comment (nmalfroy) [reply src/main/java/bio/terra/service/filedata/DrsService.java:801]: Huh, that's interesting.  lemme see if I can think of a way to handle that

File: src/test/java/bio/terra/service/snapshot/SnapshotDaoTest.java
Comment (nmalfroy) [reply src/test/java/bio/terra/service/snapshot/SnapshotDaoTest.java:843]: Oh yeah, totally!  I'm thinking I might pick that one up soon

File: src/main/java/bio/terra/service/filedata/DrsIdDao.java
Comment (nmalfroy) [reply src/main/java/bio/terra/service/filedata/DrsIdDao.java:18]: yeah, I was torn between my desire to have all constants at the top of the class and having the sql close to where it gets used.  Eh, I'm sold. I'll reorg those.

File: src/main/java/bio/terra/service/snapshot/SnapshotDao.java
Comment (nmalfroy) [reply src/main/java/bio/terra/service/snapshot/SnapshotDao.java:655]: Just wrapping the text. Not sure if that was the question?

File: src/main/java/bio/terra/service/snapshot/flight/create/SnapshotRecordFileIdsAzureStep.java
Comment (nmalfroy) [reply src/main/java/bio/terra/service/snapshot/flight/create/SnapshotRecordFileIdsAzureStep.java:17]: Yeah, I like it because it also makes logging more relevant

File: src/main/java/bio/terra/service/dataset/Dataset.java
Comment (nmalfroy) [reply src/main/java/bio/terra/service/dataset/Dataset.java]: fair enough

Comment (nmalfroy) [reply src/main/java/bio/terra/service/dataset/Dataset.java]: also, that was a whiff on my end.  It's camel cased everywhere else

File: src/main/java/bio/terra/service/filedata/flight/ingest/FileIngestWorkerFlight.java
Comment (nmalfroy) [reply src/main/java/bio/terra/service/filedata/flight/ingest/FileIngestWorkerFlight.java]: That's a good idea.  I'll see how best to do that

File: src/test/java/bio/terra/service/filedata/DrsServiceTest.java
Comment (nmalfroy) [reply src/test/java/bio/terra/service/filedata/DrsServiceTest.java]: Also good point re: new test.  I have it checking when there are invalid checksums but I'll add a check for that

File: src/main/resources/api/data-repository-openapi.yaml
Comment (nmalfroy) [reply src/main/resources/api/data-repository-openapi.yaml:3726]: Yeah, I kinda like leaving the default.  Also I wanted to keep the pattern with how we've been doing it for other parameters

File: src/main/java/bio/terra/service/filedata/flight/ingest/IngestFileAzurePrimaryDataStep.java
Comment (nmalfroy) [reply src/main/java/bio/terra/service/filedata/flight/ingest/IngestFileAzurePrimaryDataStep.java]: Ah, it actually calls the file id generation step though as I type this, I'm realizing that it shouldn't...

File: src/main/java/bio/terra/service/filedata/azure/AzureSynapsePdao.java
Comment (snf2ye) [reply src/main/java/bio/terra/service/filedata/azure/AzureSynapsePdao.java]: Definitely useful for debugging.
I'd still urge us to remove it before merging for a few reasons listed below (BUT, I kinda hope y'all can prove my concerns wrong b/c I do think it would be helpful):

- This would print out the secret that's used to create the scoped credentials. While this does expire, I would still think this is a security risk to have this printed in this less-secure location.
- This is going to produce a lot of logs($$$), especially as we move to the larger by query snapshot creation where we have to walk through each table 
- There is an alternative --  There is a "monitor" feature in synapse studio where you can search through recent queries
image

Comment (nmalfroy) [reply src/main/java/bio/terra/service/filedata/azure/AzureSynapsePdao.java]: Whoa cool!  You know, I think I'll remove for now to be safe

**Review by snf2ye — APPROVED** [review]

File: src/main/java/bio/terra/service/filedata/azure/AzureSynapsePdao.java
Comment (nmalfroy) [reply src/main/java/bio/terra/service/filedata/azure/AzureSynapsePdao.java]: Note: going to change to use debug logging for this

**Review by okotsopoulos — APPROVED** [review]

File: src/main/java/bio/terra/service/filedata/DrsIdService.java
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/filedata/DrsIdService.java:73]: I realize that this is a very small part of what you're doing here but I have to comment on the way this code is constructing and parsing versioned IDs. Right now the code is using a fluent builder and the code is passing in a version string. But `DrsId.Builder` never enforces that a specific version of ID has the fields it should have. The code is also splitting up the code that construct IDs and parses them.

I think a better approach would
- parse and assemble in a single place
- either make the builder validate the required fields based on version, or (even better IMO) not use fluent constructors and instead use factory methods like `DrsId.createV1()`, `DrsId.createV2()`

**Pull request merged**
